### PR TITLE
Generate SSH key on users' first login

### PR DESCRIPTION
**Pull request by hanwen-pcluste**
This functionality can be disabled by generate_ssh_keys_for_users

Signed-off-by: Hanwen <hanwenli@amazon.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cookbooks/aws-parallelcluster-config/recipes/directory_service.rb [hunk 1]
```diff
@@ -49,6 +49,24 @@
...
+      source 'directory_service/generate_ssh_key.sh.erb'
+      owner 'root'
+      group 'root'
+      mode '0600'
+    end
+    filter_lines 'Configure PAM postlogin script to call generate SSH key script' do
+      path generate_ssh_key_path
+      filters(
+        [
+          { before: [/^session.*/, "session     optional      pam_exec.so #{generate_ssh_key_path}", :first] },
```
Comment (tilne) [inline cookbooks/aws-parallelcluster-config/recipes/directory_service.rb]: should we log this to something in /var/log/parallelcluster? maybe something like /var/log/parallelcluster/pam_ssh_key_generator.log?

File: cookbooks/aws-parallelcluster-config/templates/default/directory_service/generate_ssh_key.sh.erb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+set -ex
+env
+if [ ! -d /home/${PAM_USER}/.ssh ] ; then
+    mkdir /home/${PAM_USER}/.ssh
+    ssh-keygen -q -t rsa -f /home/${PAM_USER}/.ssh/id_rsa -N ''
+    cat /home/${PAM_USER}/.ssh/id_rsa.pub >> /home/${PAM_USER}/.ssh/authorized_keys
+    chmod 0600 /home/${PAM_USER}/.ssh/authorized_keys
+    ssh-keyscan <%=  node['hostname'] %> > /home/${PAM_USER}/.ssh/known_hosts
+    chmod 0600 /home/${PAM_USER}/.ssh/known_hosts
+    chown ${PAM_USER}:$(id -g ${PAM_USER}) -R /home/${PAM_USER}/.ssh
```
Comment (tilne) [inline cookbooks/aws-parallelcluster-config/templates/default/directory_service/generate_ssh_key.sh.erb:11]: can you add an else case that explicitly exits with a status of 0?

File: cookbooks/aws-parallelcluster-config/recipes/directory_service.rb (hunk 1 above)
Comment (hanwen-pcluste) [reply cookbooks/aws-parallelcluster-config/recipes/directory_service.rb]: Done

File: cookbooks/aws-parallelcluster-config/templates/default/directory_service/generate_ssh_key.sh.erb (hunk 2 above)
Comment (hanwen-pcluste) [reply cookbooks/aws-parallelcluster-config/templates/default/directory_service/generate_ssh_key.sh.erb:11]: Done

File: cookbooks/aws-parallelcluster-config/recipes/directory_service.rb [hunk 3]
```diff
@@ -49,6 +49,27 @@
...
+  if node['cluster']["directory_service"]["generate_ssh_keys_for_users"] == 'true'
+    generate_ssh_key_path = "#{node['cluster']['scripts_dir']}/generate_ssh_key.sh"
+    template generate_ssh_key_path do
+      source 'directory_service/generate_ssh_key.sh.erb'
+      owner 'root'
+      group 'root'
+      mode '0600'
+    end
+    filter_lines 'Configure PAM postlogin script to call generate SSH key script' do
+      path generate_ssh_key_path
```
Comment (tilne) [inline cookbooks/aws-parallelcluster-config/recipes/directory_service.rb]: should this be set to the path to the PAM file?

File: cookbooks/aws-parallelcluster-config/recipes/directory_service.rb [hunk 4]
```diff
@@ -49,6 +49,24 @@
...
+
+  if node['cluster']["directory_service"]["generate_ssh_keys_for_users"] == 'true'
+    generate_ssh_key_path = "#{node['cluster']['scripts_dir']}/generate_ssh_key.sh"
+    template generate_ssh_key_path do
+      source 'directory_service/generate_ssh_key.sh.erb'
+      owner 'root'
+      group 'root'
+      mode '0755'
+    end
+    filter_lines 'Configure PAM postlogin script to call generate SSH key script' do
```
Comment (tilne) [inline cookbooks/aws-parallelcluster-config/recipes/directory_service.rb]: minor: name is wrong

Comment (tilne) [reply cookbooks/aws-parallelcluster-config/recipes/directory_service.rb]: should probably be 'PAM sshd config' instead of PAM postlogin script

**Review by tilne — DISMISSED** [review]

File: cookbooks/aws-parallelcluster-config/recipes/directory_service.rb (hunk 4 above)
Comment (hanwen-pcluste) [reply cookbooks/aws-parallelcluster-config/recipes/directory_service.rb]: Done

**Review by tilne — APPROVED** [review]

**Pull request merged**